### PR TITLE
validate feed keys used needs to allow periods

### PR DESCRIPTION
**Pull request by flavio-fernandes**
The recently introduced validate_feed_key [0] function
needs to account for cases when feed_id includes a
group. For example: "sign-quotes.signtext"

Fixes adafruit/Adafruit_CircuitPython_AdafruitIO#71

[0]: https://github.com/adafruit/Adafruit_CircuitPython_AdafruitIO/commit/1e3d458ccbe86dade80ce3005a41ffed5d50b864

### CODE REVIEW

**Comment by flavio-fernandes** [comment]
@brentru @ladyada : Since you have recently looked at validate_feed_key, 
please see if this is a good fix for issue #71

**Review by brentru — APPROVED** [review]
lgtm -thanks for the PR

File: adafruit_io/adafruit_io.py [hunk 1]
```diff
@@ -40,9 +40,9 @@ def validate_feed_key(feed_key):
     """
     if len(feed_key) > 128:  # validate feed key length
         raise ValueError("Feed key must be less than 128 characters.")
-    if not bool(re.match("^[a-z0-9-]+$", feed_key)):  # validate key naming scheme
+    if not bool(re.match("^[a-z0-9-\\.]+$", feed_key)):  # validate key naming scheme
```
Comment (brentru) [inline adafruit_io/adafruit_io.py:43]: @flavio-fernandes  Why did you add `\\`? IO feeds do not contain "\" characters.

Comment (brentru) [reply adafruit_io/adafruit_io.py:43]: `^[a-z0-9-.]+$` is able to match a grouped feed such as `air-quality-sensor.aqi`

Comment (xadhoom) [reply adafruit_io/adafruit_io.py:43]: Sorry for chiming in, but the regexp in the above code allows for leading/trailing dots.

Also I think that a better regexp is something like `^[a-z0-9-]+(\.[a-z0-9-]+)?$`, which matches `foo` or `foo.bar` but not `foo.` or `foo.bar.baz`. (don't know if IO feeds can have subgroups).

Comment (brentru) [reply adafruit_io/adafruit_io.py:43]: @xadhoom IO feeds can't have subgroups. like that, it'll typically look like `feed.group`. Thanks for chiming in

Comment (brentru) [reply adafruit_io/adafruit_io.py:43]: https://learn.adafruit.com/naming-things-in-adafruit-io/aside-naming-things-in-mqtt

Comment (xadhoom) [reply adafruit_io/adafruit_io.py:43]: Ok, so I read it correctly. then my regexp above is more correct, because it matches only a group and blocks leading/trailing dots .

Comment (flavio-fernandes) [reply adafruit_io/adafruit_io.py:43]: Yes. I just did not want to make it more complex than it needs to be, following the current implementation.

Comment (flavio-fernandes) [reply adafruit_io/adafruit_io.py]: @xadhoom Please see my latest change and confirm this is what you would like the change to do.
@brentru sorry for the follow up changes. I blame @xadhoom :)

File: adafruit_io/adafruit_io.py [hunk 2]
```diff
@@ -40,9 +40,11 @@ def validate_feed_key(feed_key):
     """
     if len(feed_key) > 128:  # validate feed key length
         raise ValueError("Feed key must be less than 128 characters.")
-    if not bool(re.match("^[a-z0-9-]+$", feed_key)):  # validate key naming scheme
+    if not bool(
+        re.match("^[a-z0-9-]+(\\.[a-z0-9-]+)?$", feed_key)
```
Comment (xadhoom) [inline adafruit_io/adafruit_io.py:44]: The double `\` is not needed, you're already escaping the dot with a single `\` .

A double `\\.` here means match the empty string followed by a literal dot, which does not cause any harm here but formally is not correct :)

Comment (flavio-fernandes) [reply adafruit_io/adafruit_io.py:44]: > The double `\` is not needed, you're already escaping the dot with a single `\` .
> 
> A double `\\.` here means match the empty string followed by a literal dot, which does not cause any harm here but formally is not correct :)

hmm.... I think you are not right about that. Could you give it a try? This is what I did:

```python
>>> import re
>>> feed_key = "bad."
>>> bool(re.match("^[a-z0-9-]+(\\.[a-z0-9-]+)?$", feed_key))
False
>>> feed_key = "not.bad"
>>> bool(re.match("^[a-z0-9-]+(\\.[a-z0-9-]+)?$", feed_key))
True
>>> feed_key = "is.b!ad"
>>> bool(re.match("^[a-z0-9-]+(\\.[a-z0-9-]+)?$", feed_key))
False
>>> 
```

Also, this is what I see if I do not use the extra '\':
```
$ pre-commit run --all-files
...
pylint (library code)....................................................Failed
- hook id: pylint
- exit code: 4

************* Module adafruit_io.adafruit_io
adafruit_io/adafruit_io.py:44:30: W1401: Anomalous backslash in string: '\.'. String constant might be missing an r prefix. (anomalous-backslash-in-string)
```
Ref: https://stackoverflow.com/questions/19030952/pep8-warning-on-regex-string-in-python-eclipse

File: adafruit_io/adafruit_io.py (hunk 1 above)
Comment (xadhoom) [reply adafruit_io/adafruit_io.py]: @flavio-fernandes added a small comment to the regexp, just to be a bit more pedantic :)

Also, according to https://learn.adafruit.com/naming-things-in-adafruit-io/the-two-feed-identifiers seems that the `-` is not allowed at the beginning or at the end of the string.

So if you want a more complex one: `^[a-z0-9][a-z0-9-]*((?<!-)\.(?!-)[a-z0-9-]*)?[a-z0-9]$` . This basically avoids trailing dashes and dashes "around" the dot.

See https://regex101.com/r/5Pt0U9/1

File: adafruit_io/adafruit_io.py (hunk 2 above)
Comment (xadhoom) [reply adafruit_io/adafruit_io.py:44]: Well, as I said in this context works anyway with or without the double `\` : 
```
>>> feed_key = "bad."
>>> bool(re.match("^[a-z0-9-]+(\.[a-z0-9-]+)?$", feed_key))
False
>>> feed_key = "not.bad"
>>> bool(re.match("^[a-z0-9-]+(\.[a-z0-9-]+)?$", feed_key))
True
>>> feed_key = "is.b!ad"
>>> feed_key = "is.b!ad"
>>> bool(re.match("^[a-z0-9-]+(\.[a-z0-9-]+)?$", feed_key))
False
```

To satisfy the linter, I think you can just prefix your string with an `r` : re.match(r"^[a-z0-9-]+(\.[a-z0-9-]+)?$") , or yes use the double escape (personally I think that an `r` makes the expression more readable)

Comment (flavio-fernandes) [reply adafruit_io/adafruit_io.py]: Done.

File: adafruit_io/adafruit_io.py (hunk 1 above)
Comment (flavio-fernandes) [reply adafruit_io/adafruit_io.py]: That is getting a little too complex. I _really_ rather not...

**Review by brentru — APPROVED** [review]
Changes look good, thanks for having a discussion about this.
